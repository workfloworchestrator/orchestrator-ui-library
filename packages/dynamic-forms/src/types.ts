import React, { FormEventHandler } from 'react';
import { FieldValues, useForm } from 'react-hook-form';

import { z } from 'zod';

export type TDynamicFormFieldOptionsType = 'anyOf' | 'allOf';

export enum DfFieldFormats {
    DATE = 'date',
    HIDDEN = 'hidden',
    DATETIME = 'date-time',
    SKIP = 'skip',
    TIMESTAMP = 'timestamp',

    DEFAULT = 'text',
    DROPDOWN = 'dropdown',
    BOOLFIELD = 'boolfield',
    SELECT = 'select',
    MULTISELECT = 'multiselect',
    CHECKBOX = 'checkbox',
    RADIO = 'radio',
    SWITCH = 'switch',
    DATEPICKER = 'datepicker',
    NUMBER = 'number',
    LIST = 'list',
}

export enum DfFieldTypes {
    OPTGROUP = 'optGroup',
    SKIP = 'skip',
    LONG = 'long',
    HIDDEN = 'hidden',
    LABEL = 'label',
    SUMMARY = 'summary',
    ACCEPT = 'accept',
    NULL = 'null',

    DATE = 'date',
    DATETIME = 'date-time',
    OBJECT = 'object',
    STRING = 'string',
    ARRAY = 'array',
    BOOLEAN = 'boolean',
    NUMBER = 'number',
    TIMESTAMP = 'timestamp',
    LIST = 'list',
    DEFINED_LIST = 'defined-list',
    DEFINED_MULTI_LIST = 'defined-multi-list',
    TWO_OPTION_LIST = 'two-options-list',
}

export enum DFFieldReturnType {
    STRING = 'string',
    NUMBER = 'number',
    DATETIME = 'date-time',
}
export interface IValidationErrorDetails {
    detail: string;
    source: IDynamicFormApiValidationError[];
    mapped: {
        [fieldId: string]: {
            value: string;
            msg: string;
        };
    };
}

export interface IDynamicFormsContextProps {
    isLoading: boolean;
    isSending: boolean;
    isFullFilled: boolean;
    theReactHookForm: ReturnType<typeof useForm>;
    errorDetails?: IValidationErrorDetails;
    formData?: IDynamicForm;
    debugMode?: boolean;
    title?: string;
    sendLabel?: string;
    onCancel?: () => void;
    submitForm: FormEventHandler<HTMLFormElement>;
    resetForm: (e: MouseEvent) => void;
    successNotice?: React.ReactNode;
    headerComponent?: React.ReactNode;
    footerComponent?: React.ReactNode;
}

export type DynamicFormsMetaData = {
    [key: string | number]: DfFieldValue;
};

export type DfFormProvider = ({
    formKey,
    requestBody,
}: {
    formKey: string;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    requestBody: any;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
}) => Promise<Record<string, any>>;

export type DfLabelProvider = ({
    formKey,
    id,
}: {
    formKey: string;
    id?: string | null;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
}) => Promise<Record<string, any>>;

export type DfDataProvider = () => Promise<IDynamicFormsLabels>;

export type CustomValidationRuleFn = (
    fieldConfig: IDynamicFormField,
    theReactHookForm?: ReturnType<typeof useForm>,
) => Zod.ZodTypeAny | undefined;

export interface IDynamicFormsContextConfig {
    // use custom method to provide data for the form. This overwrites data fetched from labels endpoint
    dataProvider: DfDataProvider;

    // use custom method for providing labels and data
    labelProvider: DfLabelProvider;

    // use a custom method for providing the form definition
    formProvider: DfFormProvider;

    // Extend field definitions
    fieldDetailProvider?: IFieldDefinitionProvider;

    // be able to refresh the provided data
    dataProviderCacheKey?: number;

    // whenever a fieldvalue changes, do something
    onFieldChangeHandler?: onFieldChangeHandlerFn;

    // provide custom validation rules for fields
    customValidationRules?: CustomValidationRuleFn;

    // This is a temp solution for 2 different Backend implementations of pydantic forms
    // we will remove this once the backends both work the same
    // Portal backend has a hack that is a leftover from previous implementation
    // So for only that backend, we should not send the data as a array
    tmp_pydanticFormsOriginalImplementation?: boolean;
}

export interface IDynamicFormsContextInitialProps {
    formKey: string;
    formIdKey?: string;
    title?: string;
    sendLabel?: string;
    metaData?: DynamicFormsMetaData;
    successNotice?: React.ReactNode;
    onSuccess?: (fieldValues: FieldValues) => void;
    onCancel?: () => void;
    children: (props: IDynamicFormsContextProps) => React.ReactNode;
    headerComponent?: React.ReactNode;
    footerComponent?: React.ReactNode;

    config: IDynamicFormsContextConfig;
}

export interface IFieldDefinitionProvider {
    [fieldId: string]: Partial<IDynamicFormField>;
}

export type onFieldChangeHandlerFn = (
    field: {
        name?: string;
        type?: string;
        value: DfFieldValue;
    },
    theReactHookForm: ReturnType<typeof useForm>,
) => void;

export interface IDynamicFormApiResponseDefEnum {}

export interface TJsonSchemaRef {
    $ref: string;
}

export interface TDynamicFormFieldAnyOfDef {
    items?: TJsonSchemaRef;
    type: 'null' | 'array';
}

export type ListFieldType =
    | DfFieldTypes.STRING
    | DfFieldTypes.DATE
    | DfFieldTypes.NUMBER
    | DfFieldTypes.NULL;

export interface ITDynamicFormFieldAnyOfResolvedItems {
    enum: string[];
    title: string;
    type: ListFieldType;
    maxLength?: number;
    minLength?: number;
    pattern?: string;
}

export interface TDynamicFormFieldAnyOfResolved {
    items?: ITDynamicFormFieldAnyOfResolvedItems;
    enum?: string[];
    format?: 'date' | 'date-time';
    type: ListFieldType;
    maxLength?: number;
    minLength?: number;
    pattern?: string;
}

export interface IDynamicFormApiResponseProperty {
    type: DfFieldTypes;
    anyOf?: TDynamicFormFieldAnyOfDef[];
    default?: string | null;
    title: string;
}

export interface IDynamicFormApiResponsePropertyResolved {
    type: DfFieldTypes;

    anyOf?: TDynamicFormFieldAnyOfResolved[];
    oneOf?: TDynamicFormFieldAnyOfResolved[];
    allOf?: TDynamicFormFieldAnyOfResolved[];

    items?: ITDynamicFormFieldAnyOfResolvedItems;
    enum?: string[];

    default?: string | null;
    title: string;
    format: DfFieldFormats;

    uniforms?: {
        disabled: boolean;
    };

    // validation props
    maxLength?: number;
    minLength?: number;
    pattern?: string;
}

export interface IDynamicFormApiResponseBase {
    title: string;
    description: string;
    additionalProperties: boolean;
    type: 'object';
    required?: string[];
    $defs?: {
        [definitionId: string]: {
            enum: string[];
            title: string;
            type: DfFieldTypes;
        };
    };
}

export interface IDynamicFormApiResponse extends IDynamicFormApiResponseBase {
    properties: {
        [propId: string]: IDynamicFormApiResponseProperty;
    };
}

export interface IDynamicFormApiErrorResponse {
    detail?: string;
    status: number;
    form: IDynamicFormApiResponse;
    success?: boolean;
    validation_errors: IDynamicFormApiValidationError[];
}

export interface IDynamicFormApiValidationError {
    input: string;
    loc: string[];
    msg: string;
    type: string;
    url: string; //"https://errors.pydantic.dev/2.4/v/extra_forbidden"
}

export interface IDynamicFormApiRefResolved
    extends IDynamicFormApiResponseBase {
    properties: {
        [propId: string]: IDynamicFormApiResponsePropertyResolved;
    };
}

export interface IDynamicFormsLabels {
    [key: string]: string[] | number[] | string | number | null;
}

export interface IDynamicFormsLabelResponse {
    labels: IDynamicFormsLabels;
}

export interface IDynamicFormFieldValidation {
    maxLength?: number;
    minLength?: number;
    pattern?: string;
    isNullable?: boolean;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type DfFieldValue = any;

export interface IDynamicFormFieldAttributes {
    disabled?: boolean;
}

export interface IDynamicFormField {
    id: string;
    title: string;
    description: string;
    type: DfFieldTypes;
    format: DfFieldFormats;
    options: string[];
    disabledOptions?: string[];
    default?: DfFieldValue;
    required: boolean;
    isEnumField: boolean;
    schemaField: IDynamicFormApiResponsePropertyResolved;
    validation: IDynamicFormFieldValidation;
    attributes: IDynamicFormFieldAttributes;
    validator?: FormZodValidationFn;
    FormElement?: FormElementComponent;
    matchedFieldResult?: DfFieldConfig;
}

export interface IDynamicFormFieldSection {
    id: string;
    title: string;
    fields: string[];
}

export enum IDynamicFormState {
    NEW = 'new',
    IN_PROGRESS = 'in-progress',
    FINISHED = 'finished',
}

export interface IDynamicForm {
    title: string;
    description: string;
    state: IDynamicFormState;
    fields: IDynamicFormField[];
    sections: IDynamicFormFieldSection[];
}

export type FormZodValidationFn = (field: IDynamicFormField) => z.ZodTypeAny;
export type FormElementComponent = (props: IDFInputFieldProps) => JSX.Element;

export interface FormComponent {
    Element: FormElementComponent;
    validator?: FormZodValidationFn;
}

export interface DfFieldConfig {
    id: string;
    Component: FormComponent;
    matcher?: (field: IDynamicFormField) => boolean;
    preventColRender?: boolean;
}

export type DfFieldsConfig = DfFieldConfig[];

export interface IDFInputFieldProps {
    field: IDynamicFormField;
}

export interface IDFZodValidationPresets {
    [type: string]: FormZodValidationFn;
}
