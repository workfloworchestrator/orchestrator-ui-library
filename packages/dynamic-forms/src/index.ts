import DynamicForm from "./components/form/ContextWrap"

export default DynamicForm
