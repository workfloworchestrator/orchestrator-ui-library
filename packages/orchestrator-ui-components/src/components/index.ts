export * from './WfoBadges';
export * from './WfoPageTemplate/WfoBreadcrumbs';
export * from './WfoFilterTabs';
export * from './WfoKeyValueTable';
export * from './WfoPageTemplate';
export * from './WfoSearchBar';
export * from './WfoSettingsModal';
export * from './WfoSettingsPage';
export * from './WfoStartPage';
export * from './WfoSubscription';
export * from './WfoTable';
export * from './WfoTimeline';
export * from './WfoLoading';
export * from './WfoToastsList';
export * from './WfoForms';
export * from './WfoStartTaskButtonComboBox';
