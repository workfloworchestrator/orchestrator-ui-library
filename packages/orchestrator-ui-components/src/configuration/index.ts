export * from './constants';
export * from './policy-resources';
