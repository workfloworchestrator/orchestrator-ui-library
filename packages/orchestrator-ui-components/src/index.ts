export * from './lib/components';

export * from './lib/icons';

export * from './lib/theme';

export * from './lib/utils';

export * from './lib/contexts';

export * from './lib/types';

export * from './lib/hooks';
