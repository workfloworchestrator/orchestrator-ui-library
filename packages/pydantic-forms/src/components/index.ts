export * from './forms';
export * from './formFields';
