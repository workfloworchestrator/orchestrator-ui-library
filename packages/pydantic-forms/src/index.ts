export * from './components/index';
export * from './contexts/index';
export * from './messages/index';
export * from './types/index';
