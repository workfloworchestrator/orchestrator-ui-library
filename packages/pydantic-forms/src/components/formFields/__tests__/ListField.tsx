import createContext from "lib/uniforms-surfnet/__tests__/_createContext";
import mount from "lib/uniforms-surfnet/__tests__/_mount";
import { ListAddField, ListField, ListItemField } from "lib/uniforms-surfnet/src";
/*
 * Copyright 2019-2023 SURF.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
import React from "react";

test("<ListField> - works", () => {
    const element = <ListField name="x" id="snapshot-test" />;
    const wrapper = mount(
        element,
        createContext({ x: { type: Array }, "x.$": { type: String } }, undefined, "listfield-works")
    );

    expect(wrapper.find(ListField)).toHaveLength(1);
    expect(wrapper.render()).toMatchSnapshot();
});

test("<ListField> - renders ListAddField", () => {
    const element = <ListField name="x" label="ListFieldLabel" />;
    const wrapper = mount(element, createContext({ x: { type: Array }, "x.$": { type: String } }));

    expect(wrapper.find(ListAddField)).toHaveLength(1);
    expect(wrapper.find(ListAddField).prop("name")).toBe("$");
});

test("<ListField> - renders correct label (specified)", () => {
    const element = <ListField name="x" label="ListFieldLabel" />;
    const wrapper = mount(element, createContext({ x: { type: Array }, "x.$": { type: String } }));

    expect(wrapper.find(".euiFormLabel")).toHaveLength(2);
    expect(wrapper.find(".euiFormLabel").at(0).text()).toEqual(expect.stringContaining("ListFieldLabel"));
});

test("<ListField> - renders correct numer of items with initialCount (specified)", () => {
    const element = <ListField name="x" initialCount={3} />;
    const wrapper = mount(element, createContext({ x: { type: Array }, "x.$": { type: String } }));

    expect(wrapper.find("input")).toHaveLength(3);
});

test("<ListField> - passes itemProps to its children", () => {
    const element = <ListField name="x" initialCount={3} itemProps={{ "data-xyz": 1 }} />;
    const wrapper = mount(element, createContext({ x: { type: Array }, "x.$": { type: String } }));

    expect(wrapper.find(ListItemField).first().prop("data-xyz")).toBe(1);
});

test("<ListField> - renders children (specified)", () => {
    const Child = jest.fn(() => <div />) as React.FC<any>;

    const element = (
        <ListField name="x" initialCount={2}>
            <Child />
            PlainText
        </ListField>
    );
    mount(element, createContext({ x: { type: Array }, "x.$": { type: String } }));

    expect(Child).toHaveBeenCalledTimes(2);
});

test("<ListField> - renders children with correct name (children)", () => {
    const Child = jest.fn(() => <div />) as React.FC<any>;

    const element = (
        <ListField name="x" initialCount={2}>
            <Child name="$" />
        </ListField>
    );
    const wrapper = mount(element, createContext({ x: { type: Array }, "x.$": { type: String } }));

    expect(wrapper.find(Child).at(0).prop("name")).toBe("0");
    expect(wrapper.find(Child).at(1).prop("name")).toBe("1");
});

test("<ListField> - renders children with correct values", () => {
    const element = <ListField name="x" id="snapshot-test" />;
    const wrapper = mount(
        element,
        createContext(
            { x: { type: Array }, "x.$": { type: String } },
            { model: { x: ["a", "b", "c"] } },
            "listfield-correct-values"
        )
    );

    expect(wrapper.find(ListItemField)).toHaveLength(3);

    expect(wrapper.find("input").at(0).prop("value")).toBe("a");
    expect(wrapper.find("input").at(1).prop("value")).toBe("b");

    expect(wrapper.find("input").at(2).prop("value")).toBe("c");
    expect(wrapper.render()).toMatchSnapshot();
});

test("<ListField> - renders children with correct name (value)", () => {
    const element = <ListField name="x" initialCount={2} id="snapshot-test" />;
    const wrapper = mount(
        element,
        createContext({ x: { type: Array }, "x.$": { type: String } }, undefined, "listfield-correct-name")
    );

    expect(wrapper.find(ListItemField).at(0).prop("name")).toBe("0");
    expect(wrapper.find(ListItemField).at(1).prop("name")).toBe("1");
    expect(wrapper.render()).toMatchSnapshot();
});

test("<ListField> - renders correctly when child is list", () => {
    const element = <ListField name="x" id="snapshot-test" />;
    const wrapper = mount(
        element,
        createContext(
            { x: { type: Array }, "x.$": { type: Array }, "x.$.$": { type: String } },
            { model: { x: [["test"]] } },
            "listfield-childlist"
        )
    );

    expect(wrapper.find(ListItemField)).toHaveLength(2);
    expect(wrapper.find(ListItemField).at(0).prop("outerList")).toBe(true);
    expect(wrapper.find(ListItemField).at(1).prop("outerList")).toBe(false);
    expect(wrapper.find(ListAddField)).toHaveLength(2);
    expect(wrapper.find(ListAddField).at(0).prop("outerList")).toBe(false);
    expect(wrapper.find(ListAddField).at(1).prop("outerList")).toBe(true);

    expect(wrapper.find("section")).toHaveLength(3);
    expect(wrapper.find("section").at(0).prop("className")).toBe("list-field outer-list");
    expect(wrapper.find("section").at(1).prop("className")).toBe("list-field");
    expect(wrapper.find(ListField).at(0).render()).toMatchSnapshot();
});

test("<ListField> - renders proper number of optional values after add new value (with initialCount)", () => {
    const element = <ListField name="x" initialCount={3} label="ListFieldLabel" />;
    const onChange = jest.fn();
    const wrapper = mount(
        element,
        createContext({ x: { type: Array, optional: true }, "x.$": { type: String } }, { onChange })
    );
    expect(wrapper.find(ListAddField).simulate("click")).toBeTruthy();
    expect(onChange).toHaveBeenNthCalledWith(1, "x", [undefined, undefined, undefined, undefined]);
});
