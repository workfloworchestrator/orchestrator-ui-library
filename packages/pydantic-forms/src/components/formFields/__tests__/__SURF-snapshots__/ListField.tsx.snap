// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListField> - renders children with correct name (value) 1`] = `
LoadedCheerio {
  "0": <div
    class="euiFlexItem css-ys1ii3-EuiFlexItem"
  >
    <section
      class="list-field"
      id="snapshot-test"
      required=""
    >
      <div
        class="euiFormRow euiFormRow--fullWidth"
        id="formrow-x-row"
      >
        <div
          class="euiFormRow__labelWrapper"
        >
          <label
            class="euiFormLabel euiFormRow__label"
            for="formrow-x"
            id="formrow-x-label"
          />
           
          <div
            class="euiText euiText--medium"
          />
        </div>
        <div
          class="euiFormRow__fieldWrapper"
        />
      </div>
      <ul>
        <li>
          <section>
            <div
              class="euiFormRow euiFormRow--fullWidth"
              id="listfield-correct-name-0003-row"
            >
              <div
                class="euiFormRow__labelWrapper"
              >
                <label
                  class="euiFormLabel euiFormRow__label"
                  for="listfield-correct-name-0003"
                  id="listfield-correct-name-0003-label"
                />
                 
                <div
                  class="euiText euiText--medium"
                />
              </div>
              <div
                class="euiFormRow__fieldWrapper"
              >
                <div
                  class="euiFormControlLayout euiFormControlLayout--fullWidth"
                >
                  <div
                    class="euiFormControlLayout__childrenWrapper"
                  >
                    <input
                      class="euiFieldText euiFieldText--fullWidth"
                      id="listfield-correct-name-0003"
                      name="x.0"
                      placeholder=""
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
            </div>
          </section>
          <div
            class="del-item"
            id="listfield-correct-name-0004.remove"
            role="button"
            tabindex="0"
          >
            <i
              class="fa fa-minus "
            />
            <label />
          </div>
        </li>
        <li>
          <section>
            <div
              class="euiFormRow euiFormRow--fullWidth"
              id="listfield-correct-name-0007-row"
            >
              <div
                class="euiFormRow__labelWrapper"
              >
                <label
                  class="euiFormLabel euiFormRow__label"
                  for="listfield-correct-name-0007"
                  id="listfield-correct-name-0007-label"
                />
                 
                <div
                  class="euiText euiText--medium"
                />
              </div>
              <div
                class="euiFormRow__fieldWrapper"
              >
                <div
                  class="euiFormControlLayout euiFormControlLayout--fullWidth"
                >
                  <div
                    class="euiFormControlLayout__childrenWrapper"
                  >
                    <input
                      class="euiFieldText euiFieldText--fullWidth"
                      id="listfield-correct-name-0007"
                      name="x.1"
                      placeholder=""
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
            </div>
          </section>
          <div
            class="del-item"
            id="listfield-correct-name-0008.remove"
            role="button"
            tabindex="0"
          >
            <i
              class="fa fa-minus "
            />
            <label />
          </div>
        </li>
        <div
          class="add-item"
          id="listfield-correct-name-000a"
          role="button"
          tabindex="0"
        >
          <i
            class="fa fa-plus "
          />
          <label />
        </div>
      </ul>
    </section>
  </div>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`<ListField> - renders children with correct values 1`] = `
LoadedCheerio {
  "0": <div
    class="euiFlexItem css-ys1ii3-EuiFlexItem"
  >
    <section
      class="list-field"
      id="snapshot-test"
      required=""
    >
      <div
        class="euiFormRow euiFormRow--fullWidth"
        id="formrow-x-row"
      >
        <div
          class="euiFormRow__labelWrapper"
        >
          <label
            class="euiFormLabel euiFormRow__label"
            for="formrow-x"
            id="formrow-x-label"
          />
           
          <div
            class="euiText euiText--medium"
          />
        </div>
        <div
          class="euiFormRow__fieldWrapper"
        />
      </div>
      <ul>
        <li>
          <section>
            <div
              class="euiFormRow euiFormRow--fullWidth"
              id="listfield-correct-values-0003-row"
            >
              <div
                class="euiFormRow__labelWrapper"
              >
                <label
                  class="euiFormLabel euiFormRow__label"
                  for="listfield-correct-values-0003"
                  id="listfield-correct-values-0003-label"
                />
                 
                <div
                  class="euiText euiText--medium"
                />
              </div>
              <div
                class="euiFormRow__fieldWrapper"
              >
                <div
                  class="euiFormControlLayout euiFormControlLayout--fullWidth"
                >
                  <div
                    class="euiFormControlLayout__childrenWrapper"
                  >
                    <input
                      class="euiFieldText euiFieldText--fullWidth"
                      id="listfield-correct-values-0003"
                      name="x.0"
                      placeholder=""
                      type="text"
                      value="a"
                    />
                  </div>
                </div>
              </div>
            </div>
          </section>
          <div
            class="del-item"
            id="listfield-correct-values-0004.remove"
            role="button"
            tabindex="0"
          >
            <i
              class="fa fa-minus "
            />
            <label />
          </div>
        </li>
        <li>
          <section>
            <div
              class="euiFormRow euiFormRow--fullWidth"
              id="listfield-correct-values-0007-row"
            >
              <div
                class="euiFormRow__labelWrapper"
              >
                <label
                  class="euiFormLabel euiFormRow__label"
                  for="listfield-correct-values-0007"
                  id="listfield-correct-values-0007-label"
                />
                 
                <div
                  class="euiText euiText--medium"
                />
              </div>
              <div
                class="euiFormRow__fieldWrapper"
              >
                <div
                  class="euiFormControlLayout euiFormControlLayout--fullWidth"
                >
                  <div
                    class="euiFormControlLayout__childrenWrapper"
                  >
                    <input
                      class="euiFieldText euiFieldText--fullWidth"
                      id="listfield-correct-values-0007"
                      name="x.1"
                      placeholder=""
                      type="text"
                      value="b"
                    />
                  </div>
                </div>
              </div>
            </div>
          </section>
          <div
            class="del-item"
            id="listfield-correct-values-0008.remove"
            role="button"
            tabindex="0"
          >
            <i
              class="fa fa-minus "
            />
            <label />
          </div>
        </li>
        <li>
          <section>
            <div
              class="euiFormRow euiFormRow--fullWidth"
              id="listfield-correct-values-000b-row"
            >
              <div
                class="euiFormRow__labelWrapper"
              >
                <label
                  class="euiFormLabel euiFormRow__label"
                  for="listfield-correct-values-000b"
                  id="listfield-correct-values-000b-label"
                />
                 
                <div
                  class="euiText euiText--medium"
                />
              </div>
              <div
                class="euiFormRow__fieldWrapper"
              >
                <div
                  class="euiFormControlLayout euiFormControlLayout--fullWidth"
                >
                  <div
                    class="euiFormControlLayout__childrenWrapper"
                  >
                    <input
                      class="euiFieldText euiFieldText--fullWidth"
                      id="listfield-correct-values-000b"
                      name="x.2"
                      placeholder=""
                      type="text"
                      value="c"
                    />
                  </div>
                </div>
              </div>
            </div>
          </section>
          <div
            class="del-item"
            id="listfield-correct-values-000c.remove"
            role="button"
            tabindex="0"
          >
            <i
              class="fa fa-minus "
            />
            <label />
          </div>
        </li>
        <div
          class="add-item"
          id="listfield-correct-values-000e"
          role="button"
          tabindex="0"
        >
          <i
            class="fa fa-plus "
          />
          <label />
        </div>
      </ul>
    </section>
  </div>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`<ListField> - renders correctly when child is list 1`] = `
LoadedCheerio {
  "0": <div
    class="euiFlexItem css-ys1ii3-EuiFlexItem"
  >
    <section
      class="list-field outer-list"
      id="snapshot-test"
      required=""
    >
      <div
        class="euiFormRow euiFormRow--fullWidth"
        id="formrow-x-row"
      >
        <div
          class="euiFormRow__labelWrapper"
        >
          <label
            class="euiFormLabel euiFormRow__label"
            for="formrow-x"
            id="formrow-x-label"
          />
           
          <div
            class="euiText euiText--medium"
          />
        </div>
        <div
          class="euiFormRow__fieldWrapper"
        />
      </div>
      <ul>
        <li>
          <div
            class="euiFlexItem css-ys1ii3-EuiFlexItem"
          >
            <section
              class="list-field"
              id="listfield-childlist-0004"
            >
              <div
                class="euiFormRow euiFormRow--fullWidth"
                id="formrow-x.0-row"
              >
                <div
                  class="euiFormRow__labelWrapper"
                >
                  <label
                    class="euiFormLabel euiFormRow__label"
                    for="formrow-x.0"
                    id="formrow-x.0-label"
                  />
                   
                  <div
                    class="euiText euiText--medium"
                  />
                </div>
                <div
                  class="euiFormRow__fieldWrapper"
                />
              </div>
              <ul>
                <li>
                  <section>
                    <div
                      class="euiFormRow euiFormRow--fullWidth"
                      id="listfield-childlist-0007-row"
                    >
                      <div
                        class="euiFormRow__labelWrapper"
                      >
                        <label
                          class="euiFormLabel euiFormRow__label"
                          for="listfield-childlist-0007"
                          id="listfield-childlist-0007-label"
                        />
                         
                        <div
                          class="euiText euiText--medium"
                        />
                      </div>
                      <div
                        class="euiFormRow__fieldWrapper"
                      >
                        <div
                          class="euiFormControlLayout euiFormControlLayout--fullWidth"
                        >
                          <div
                            class="euiFormControlLayout__childrenWrapper"
                          >
                            <input
                              class="euiFieldText euiFieldText--fullWidth"
                              id="listfield-childlist-0007"
                              name="x.0.0"
                              placeholder=""
                              type="text"
                              value="test"
                            />
                          </div>
                        </div>
                      </div>
                    </div>
                  </section>
                  <div
                    class="del-item"
                    id="listfield-childlist-0008.remove"
                    role="button"
                    tabindex="0"
                  >
                    <i
                      class="fa fa-minus "
                    />
                    <label />
                  </div>
                </li>
                <div
                  class="add-item"
                  id="listfield-childlist-000a"
                  role="button"
                  tabindex="0"
                >
                  <i
                    class="fa fa-plus "
                  />
                  <label />
                </div>
              </ul>
            </section>
          </div>
          <div
            class="del-item"
            id="listfield-childlist-000c.remove"
            role="button"
            tabindex="0"
          >
            <i
              class="fa fa-minus "
            />
            <label>
              forms.fields.x_del
            </label>
          </div>
        </li>
        <div
          class="add-item"
          id="listfield-childlist-000e"
          role="button"
          tabindex="0"
        >
          <i
            class="fa fa-plus "
          />
          <label>
            forms.fields.x_add
          </label>
        </div>
      </ul>
    </section>
  </div>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`<ListField> - works 1`] = `
LoadedCheerio {
  "0": <div
    class="euiFlexItem css-ys1ii3-EuiFlexItem"
  >
    <section
      class="list-field"
      id="snapshot-test"
      required=""
    >
      <div
        class="euiFormRow euiFormRow--fullWidth"
        id="formrow-x-row"
      >
        <div
          class="euiFormRow__labelWrapper"
        >
          <label
            class="euiFormLabel euiFormRow__label"
            for="formrow-x"
            id="formrow-x-label"
          />
           
          <div
            class="euiText euiText--medium"
          />
        </div>
        <div
          class="euiFormRow__fieldWrapper"
        />
      </div>
      <ul>
        <li>
          <section>
            <div
              class="euiFormRow euiFormRow--fullWidth"
              id="listfield-works-0003-row"
            >
              <div
                class="euiFormRow__labelWrapper"
              >
                <label
                  class="euiFormLabel euiFormRow__label"
                  for="listfield-works-0003"
                  id="listfield-works-0003-label"
                />
                 
                <div
                  class="euiText euiText--medium"
                />
              </div>
              <div
                class="euiFormRow__fieldWrapper"
              >
                <div
                  class="euiFormControlLayout euiFormControlLayout--fullWidth"
                >
                  <div
                    class="euiFormControlLayout__childrenWrapper"
                  >
                    <input
                      class="euiFieldText euiFieldText--fullWidth"
                      id="listfield-works-0003"
                      name="x.0"
                      placeholder=""
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
            </div>
          </section>
          <div
            class="del-item"
            id="listfield-works-0004.remove"
            role="button"
            tabindex="0"
          >
            <i
              class="fa fa-minus "
            />
            <label />
          </div>
        </li>
        <div
          class="add-item"
          id="listfield-works-0006"
          role="button"
          tabindex="0"
        >
          <i
            class="fa fa-plus "
          />
          <label />
        </div>
      </ul>
    </section>
  </div>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
